src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+
+            // Acquire role name
+
+            var roleRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/"),
+                Method = HttpMethod.Get
+            };
+            roleRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var roleName = await GetHttpContent(roleRequest, "Failed to acquire EC2 role name.").ConfigureAwait(false);
+
+            // Acquire credentials
+
+            var credentialsUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/");
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(credentialsUri, roleName),
+                Method = HttpMethod.Get
+            };
+            credentialsRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var credentials = await GetHttpContent(credentialsRequest, "Failed to acquire EC2 credentials.").ConfigureAwait(false);
+
+            return credentials;
+        }
+
+        private static async Task<string> GetECSResponse(Uri baseUri, string relativeUri)
+        {
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, relativeUri),
+                Method = HttpMethod.Get
+            };
+
+            return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+        }
+
+        private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+        {
+            HttpResponseMessage response;
+            try
+            {
+                response = await _httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new MongoClientException(exceptionMessage);
+            }
+
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private static void ValidateCredentials(string username, string password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        private static void ValidateCredentials(string username, SecureString password, string sessionToken)
```

Comment:
I may not fully understand what `MongoUtils.ToInsecureString` is supposed to be used for but why not just use that? E.g., replace the duplicated logic in this method with 
``` 
private static void ValidateCredentials(string username, SecureString password, string sessionToken)
{
    return ValidateCredentials(username, MongoUtils.ToInsecureString(password), sessionToken)
}
```

Reply:
I believe a tiny bit of duplication here is fine as we want to defer decrypting a SecureString as long as possible to minimize the time that the password hangs out in memory in clear text.

If we want to minimize the code duplication, perhaps a method could be created with that takes in an `object password` that also throws an exception if the password is not of type string or SecureString?